lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs

```diff
@@ -0,0 +1,148 @@
+import Cardano.Wallet.Network
+    ( ErrNetworkUnavailable, NetworkLayer (..) )
+import Cardano.Wallet.Primitive.Types
+    ( Coin (..), GenesisParameters (..), PoolId )
+import Cardano.Wallet.Shelley.Compatibility
+    ( Shelley, toPoint )
+import Cardano.Wallet.Shelley.Network
+    ( NodePoolLsqData (..) )
+import Cardano.Wallet.Unsafe
+    ( unsafeMkPercentage, unsafeRunExceptT )
+import Control.Monad.IO.Class
+    ( liftIO )
+import Control.Monad.Trans.Except
+    ( ExceptT (..), runExceptT )
+import Data.Map
+    ( Map )
+import Data.Map.Merge.Strict
+    ( dropMissing, traverseMissing, zipWithMatched )
+import Data.Ord
+    ( Down (..) )
+import Data.Quantity
+    ( Percentage (..), Quantity (..) )
+import Data.Sort
+    ( sortOn )
+import Data.Word
+    ( Word64 )
+import GHC.Generics
+    ( Generic )
+
+import qualified Cardano.Wallet.Api.Types as Api
+import qualified Data.Map.Merge.Strict as Map
+import qualified Data.Map.Strict as Map
+
+-- | Stake Pool Data fields fetched from the node via LSQ
+data PoolLsqMetrics = PoolLsqMetrics
+    { nonMyopicMemberRewards :: Quantity "lovelace" Word64
+    , relativeStake :: Percentage
+    , saturation :: Double
+    } deriving (Eq, Show, Generic)
+
+combineLsqData
+    :: NodePoolLsqData
+    -> Map PoolId PoolLsqMetrics
+combineLsqData NodePoolLsqData{nOpt, rewards, stake} =
+    Map.merge stakeButNoRewards rewardsButNoStake bothPresent stake rewards
+  where
+    -- calculate the saturation from the relative stake
+    sat s = fromRational $ (getPercentage s) / (1 / fromIntegral nOpt)
+
+    -- If we fetch non-myopic member rewards of pools using the wallet
+    -- balance of 0, the resulting map will be empty. So we set the rewards
+    -- to 0 here:
+    stakeButNoRewards = traverseMissing $ \_k s -> pure $ PoolLsqMetrics
+        { nonMyopicMemberRewards = Quantity 0
+        , relativeStake = s
+        , saturation = (sat s)
+        }
+
+    rewardsButNoStake = dropMissing
+
+    bothPresent       = zipWithMatched  $ \_k s r -> PoolLsqMetrics r s (sat s)
+
+readBlockProductions :: IO (Map PoolId Int)
+readBlockProductions = return Map.empty
+
+--
+-- Api Server Handler
+--
+
+data StakePoolLayer = StakePoolLayer
+    { knownPools :: IO [PoolId]
+    , listStakePools :: Coin -> ExceptT ErrNetworkUnavailable IO [Api.ApiStakePool]
+    }
+
+newStakePoolLayer
+    :: GenesisParameters
+    -> NetworkLayer IO (IO Shelley) b
+    -> StakePoolLayer
+newStakePoolLayer gp nl = StakePoolLayer
+    { knownPools = _knownPools
+    , listStakePools = _listPools
+    }
+  where
+    dummyCoin = Coin 0
+
+    -- Note: We shouldn't have to do this conversion.
+    el = getEpochLength gp
+    gh = getGenesisBlockHash gp
+    getTip = fmap (toPoint gh el) . liftIO $ unsafeRunExceptT $ currentNodeTip nl
+
+    _knownPools
+        :: IO [PoolId]
+    _knownPools = do
+        pt <- getTip
+        res <- runExceptT $ map fst . Map.toList
+            . combineLsqData <$> stakeDistribution nl pt dummyCoin
```

Comment:
This shouldn't use the network layer but use the database. Yet, you'll need the work on my branch for that. I'll modify this function in there.

Reply:
👍 Though why?

I know `knownPools` are used to validate the users's choice of delegation.

The pool worker is less likely to be in sync with the network than the node.

Wouldn't implementing it in terms of the LSQ make more sense?